opentelemetry-sdk/tests/trace/test_span_processor.py

```diff
@@ -0,0 +1,314 @@
+
+        # Span processors are created but not added to the tracer yet
+        sp = MySpanProcessor("SP1", spans_calls_list)
+
+        with tracer.start_as_current_span("foo"):
+            with tracer.start_as_current_span("bar"):
+                with tracer.start_as_current_span("baz"):
+                    # add span processor after spans have been created
+                    tracer_provider.add_span_processor(sp)
+
+                expected_list.append(span_event_end_fmt("SP1", "baz"))
+
+            expected_list.append(span_event_end_fmt("SP1", "bar"))
+
+        expected_list.append(span_event_end_fmt("SP1", "foo"))
+
+        self.assertListEqual(spans_calls_list, expected_list)
+
+
+class TestSynchronousMultiSpanProcessor(unittest.TestCase):
+    def test_on_start(self):
+        multi_processor = trace.SynchronousMultiSpanProcessor()
+
+        mocks = [mock.Mock(spec=trace.SpanProcessor) for _ in range(0, 5)]
+        for mock_processor in mocks:
+            multi_processor.add_span_processor(mock_processor)
+
+        multi_processor.on_start(mock.Mock(spec=trace.Span))
+
+        for mock_processor in mocks:
+            self.assertEqual(1, mock_processor.on_start.call_count)
+
+    def test_on_end(self):
+        multi_processor = trace.SynchronousMultiSpanProcessor()
+
+        mocks = [mock.Mock(spec=trace.SpanProcessor) for _ in range(0, 5)]
+        for mock_processor in mocks:
+            multi_processor.add_span_processor(mock_processor)
+
+        multi_processor.on_end(mock.Mock(spec=trace.Span))
+
+        for mock_processor in mocks:
+            self.assertEqual(1, mock_processor.on_end.call_count)
+
+    def test_on_shutdown(self):
+        multi_processor = trace.SynchronousMultiSpanProcessor()
+
+        mocks = [mock.Mock(spec=trace.SpanProcessor) for _ in range(0, 5)]
+        for mock_processor in mocks:
+            multi_processor.add_span_processor(mock_processor)
+
+        multi_processor.shutdown()
+
+        for mock_processor in mocks:
+            self.assertEqual(1, mock_processor.shutdown.call_count)
+
+    def test_force_flush(self):
+        multi_processor = trace.SynchronousMultiSpanProcessor()
+
+        mocks = [mock.Mock(spec=trace.SpanProcessor) for _ in range(0, 5)]
+        for mock_processor in mocks:
+            multi_processor.add_span_processor(mock_processor)
+        timeout_millis = 100
+
+        flushed = multi_processor.force_flush(timeout_millis)
+
+        self.assertTrue(flushed)
+        for mock_processor in mocks:
+            self.assertEqual(1, mock_processor.force_flush.call_count)
+
+    def test_force_flush_late_by_timeout(self):
+        multi_processor = trace.SynchronousMultiSpanProcessor()
+
+        mock_processor1 = mock.Mock(spec=trace.SpanProcessor)
+        multi_processor.add_span_processor(mock_processor1)
+        mock_processor2 = mock.Mock(spec=trace.SpanProcessor)
+        multi_processor.add_span_processor(mock_processor2)
+
+        with (
```

Comment:
this test is fairly implementation dependent. That works here but is brittle to the implementation details of the code. You could test this by creating a SpanProcessor that uses time.sleep for a specified period (maybe 50ms), and then having a SpanProcessor configured with an aggressive timeout.